### PR TITLE
Skip aggregate functions on schema dumping (#34)

### LINKED ISSUE #34: Schema dump fails in postgis
https://github.com/teoljungberg/fx/blob/53f8b3fb651e7574f98f6ff3685361a13de0981c/lib/fx/adapters/postgres/functions.rb#L11

This throws the following error if you have aggregate functions (Postgis has a lot of these):

```
ActiveRecord::StatementInvalid: PG::WrongObjectType: ERROR:  "array_accum" is an aggregate function
:           SELECT
              pp.proname AS name,
              pg_get_functiondef(pp.oid) AS definition
          FROM pg_proc pp
          JOIN pg_namespace pn
              ON pn.oid = pp.pronamespace
          LEFT JOIN pg_depend pd
              ON pd.objid = pp.oid AND pd.deptype = 'e'
          WHERE pn.nspname = 'public' AND pd.objid IS NULL
          ORDER BY pp.oid;
```

Stacktrace if that helps:

```
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:73:in `async_exec'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:73:in `block (2 levels) in execute'
/bundle/gems/activesupport-5.1.6.2/lib/active_support/dependencies/interlock.rb:46:in `block in permit_concurrent_loads'
/bundle/gems/activesupport-5.1.6.2/lib/active_support/concurrency/share_lock.rb:185:in `yield_shares'
/bundle/gems/activesupport-5.1.6.2/lib/active_support/dependencies/interlock.rb:45:in `permit_concurrent_loads'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:72:in `block in execute'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/abstract_adapter.rb:613:in `block (2 levels) in log'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/abstract_adapter.rb:612:in `block in log'
/bundle/gems/activesupport-5.1.6.2/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/abstract_adapter.rb:604:in `log'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:71:in `execute'
/bundle/gems/fx-0.5.0/lib/fx/adapters/postgres/functions.rb:47:in `functions_from_postgres'
/bundle/gems/fx-0.5.0/lib/fx/adapters/postgres/functions.rb:39:in `all'
/bundle/gems/fx-0.5.0/lib/fx/adapters/postgres/functions.rb:28:in `all'
/bundle/gems/fx-0.5.0/lib/fx/adapters/postgres.rb:50:in `functions'
/bundle/gems/fx-0.5.0/lib/fx/schema_dumper/function.rb:25:in `dumpable_functions_in_database'
/bundle/gems/fx-0.5.0/lib/fx/schema_dumper/function.rb:13:in `functions'
/bundle/gems/fx-0.5.0/lib/fx/schema_dumper/function.rb:9:in `tables'
/bundle/gems/fx-0.5.0/lib/fx/schema_dumper/trigger.rb:8:in `tables'
/bundle/gems/scenic-1.5.1/lib/scenic/schema_dumper.rb:7:in `tables'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/schema_dumper.rb:37:in `dump'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/schema_dumper.rb:21:in `dump'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/railties/databases.rake:241:in `block (4 levels) in <top (required)>'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/railties/databases.rake:240:in `open'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/railties/databases.rake:240:in `block (3 levels) in <top (required)>'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/railties/databases.rake:66:in `block (2 levels) in <top (required)>'
/bundle/gems/activerecord-5.1.6.2/lib/active_record/railties/databases.rake:59:in `block (2 levels) in <top (required)>'
/bundle/gems/railties-5.1.6.2/lib/rails/commands/rake/rake_command.rb:21:in `block in perform'
/bundle/gems/railties-5.1.6.2/lib/rails/commands/rake/rake_command.rb:18:in `perform'
/bundle/gems/railties-5.1.6.2/lib/rails/command.rb:46:in `invoke'
/bundle/gems/railties-5.1.6.2/lib/rails/commands.rb:16:in `<top (required)>'
bin/rails:4:in `require'
bin/rails:4:in `<main>'
```

Basically `pg_get_functiondef()` fails for aggregate functions.

### PR DESCRIPTION
**Pull request by vladimirtemnikov**
This should resolve #34 
Postgres can't return definitions for aggregate functions, so we need to exclude them from query. 

In addition the way to do this varies by Postgres version.
Accordingly to PG 11 changelog:

> Replace system catalog pg_proc's proisagg and proiswindow columns with prokind (Peter Eisentraut)
> This new column more clearly distinguishes functions, procedures, aggregates, and window functions.

it's possible by adding condition `pg_proc.prokind != 'a'` for Postgres >= 11 version or `pg_proc.proisagg = 'f'` for other versions.

### CODE REVIEW

**Comment by vladimirtemnikov** [comment]
Something wrong with pg on travis

**Comment by teoljungberg** [comment]
@vladimirtemnikov Can you expand on the changes in the PR description? I like PRs, and later commits to give context to the future readers of the code.

File: .travis.yml [hunk 1]
```diff
@@ -9,9 +9,6 @@ before_install:
   - "echo 'gem: --no-document' > ~/.gemrc"
   - git config --global user.name "Travis CI"
   - git config --global user.email "travis-ci@example.com"
-env:
```
Comment (teoljungberg) [inline .travis.yml]: Can this be done in a separate PR?

File: lib/fx/adapters/postgres/functions.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@ class Postgres
       class Functions
         # The SQL query used by F(x) to retrieve the functions considered
         # dumpable into `db/schema.rb`.
-        FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-EOS.freeze
+        PG10_FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-SQL.squish.freeze
```
Comment (teoljungberg) [inline lib/fx/adapters/postgres/functions.rb]: Thank you for using `SQL` in the heredoc!

File: lib/fx/adapters/postgres/functions.rb [hunk 3]
```diff
@@ -17,9 +17,22 @@ class Functions
               ON pn.oid = pp.pronamespace
           LEFT JOIN pg_depend pd
               ON pd.objid = pp.oid AND pd.deptype = 'e'
-          WHERE pn.nspname = 'public' AND pd.objid IS NULL
+          WHERE pn.nspname = 'public' AND pd.objid IS NULL AND pp.proisagg = 'f'
           ORDER BY pp.oid;
-        EOS
+        SQL
+
+        PG11_FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-SQL.squish.freeze
```
Comment (teoljungberg) [inline lib/fx/adapters/postgres/functions.rb]: I'd ideally like to have 1 query for both PG10 and 11. Is that doable?

**Comment by vladimirtemnikov** [comment]
@teoljungberg Sure :)

File: .travis.yml (hunk 1 above)
Comment (vladimirtemnikov) [reply .travis.yml]: No problem. But I guess this change is temporarily needed because I tried to follow travis's documentation and looks like they are having some kind of bug right now.

File: lib/fx/adapters/postgres/functions.rb (hunk 2 above)
Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: Nicer syntax highlighting is always cool :)

File: lib/fx/adapters/postgres/functions.rb (hunk 3 above)
Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: I think there are 2 possible ways:

1. Move this to interpolated string instead of constant.
2. Keep main part of query in constant and concat it with conditional part dynamically.

Comment (teoljungberg) [reply lib/fx/adapters/postgres/functions.rb]: Neither of those sound like a good idea to me.

I'll re-iterate. Is it possible to have 1 query that returns the functions with their definitions on Postgres 10 and 11?

File: .travis.yml (hunk 1 above)
Comment (vladimirtemnikov) [reply .travis.yml]: @teoljungberg https://github.com/teoljungberg/fx/pull/41

File: lib/fx/adapters/postgres/functions.rb (hunk 3 above)
Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: I added temp function, now it's possible.

**Comment by vladimirtemnikov** [comment]
@teoljungberg Thank you for feedback, guess I took into account all the comments.

File: lib/fx/adapters/postgres/functions.rb (hunk 3 above)
Comment (teoljungberg) [reply lib/fx/adapters/postgres/functions.rb]: This still doesn't feel right, and I don't feel that my initial question has been answered.

Is there a way to get the functions, with their definitions, that's portable between PG 10 and 11? If it isn't, I don't mind having two functions. But, ideally I'd like there to be only one. But - if that's possible, so be it.

Having a functions feels even more awkward.

Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: @teoljungberg Ah, no, it's impossible.

Comment (teoljungberg) [reply lib/fx/adapters/postgres/functions.rb]: Is there any documentation or a changelog between PG 10 and 11 that supports this claim?

Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: https://www.postgresql.org/docs/release/11.0/

> Replace system catalog pg_proc's proisagg and proiswindow columns with prokind (Peter Eisentraut)

This new column more clearly distinguishes functions, procedures, aggregates, and window functions.

Comment (teoljungberg) [reply lib/fx/adapters/postgres/functions.rb]: Good to know!

If you can document this in the PR description, and split this up to use two different queries I think we're good here.

Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: Sure, I'll make changes soon

Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: @teoljungberg Done :)

**Comment by teoljungberg** [comment]
When you force push, it makes my review work a lot more difficult. Try to push small commits and then at the end force push once we've ironed out the kinks.

**Comment by vladimirtemnikov** [comment]
I see, no problem!

File: lib/fx/adapters/postgres/functions.rb [hunk 4]
```diff
@@ -8,18 +8,37 @@ class Postgres
       class Functions
         # The SQL query used by F(x) to retrieve the functions considered
         # dumpable into `db/schema.rb`.
-        FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-EOS.freeze
+        PG_11_FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-SQL.squish.freeze
           SELECT
-              pp.proname AS name,
```
Comment (teoljungberg) [inline lib/fx/adapters/postgres/functions.rb]: Since we are reformatting these SQL calls, we can use pg_format(1):

```
SELECT
  pp.proname,
  pg_get_functiondef(pp.oid)
FROM
  pg_proc pp
  JOIN pg_namespace pn ON pn.oid = pp.pronamespace
  LEFT JOIN pg_depend pd ON pd.objid = pp.oid
    AND pd.deptype = 'e'
WHERE
  pn.nspname = 'public'
  AND pd.objid IS NULL
  AND pp.prokind != 'a'
ORDER BY
  pp.oid
```

File: lib/fx/adapters/postgres/functions.rb [hunk 5]
```diff
@@ -8,18 +8,37 @@ class Postgres
...
+            ON pd.objid = pp.oid AND pd.deptype = 'e'
+          WHERE
+            pn.nspname = 'public' AND
+            pd.objid IS NULL AND
+            pp.prokind != 'a'
+          ORDER BY pp.oid
+        SQL
+
+        PG_10_FUNCTIONS_WITH_DEFINITIONS_QUERY = <<-SQL.squish.freeze
+          SELECT
```
Comment (teoljungberg) [inline lib/fx/adapters/postgres/functions.rb]: Same reformatting with pg_format(1).

```
SELECT
  pp.proname,
  pg_get_functiondef(pp.oid)
FROM
  pg_proc pp
  JOIN pg_namespace pn ON pn.oid = pp.pronamespace
  LEFT JOIN pg_depend pd ON pd.objid = pp.oid
    AND pd.deptype = 'e'
WHERE
  pn.nspname = 'public'
  AND pd.objid IS NULL
  AND pp.proisagg = 'f'
ORDER BY
  pp.oid
```

File: lib/fx/adapters/postgres/functions.rb (hunk 4 above)
Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: Looks much better, thank you!

File: lib/fx/adapters/postgres/functions.rb (hunk 5 above)
Comment (vladimirtemnikov) [reply lib/fx/adapters/postgres/functions.rb]: Same changes done here.

**Comment by vladimirtemnikov** [comment]
Looks like everything is ok

**Comment by teoljungberg** [comment]
@vladimirtemnikov I'm terribly sorry for the slow responses, life's been busy.

I'd be happy to merge this. Can you provide me with tests that this works on PG10 and PG11? As well as updating the PR description and we're ready for merge.

**Comment by teoljungberg** [comment]
@vladimirtemnikov I'm terribly sorry for the slow responses, life's been busy.

I'd be happy to merge this. Can you provide me with tests that this works on PG10 and PG11? As well as updating the PR description and we're ready for merge.

**Comment by vladimirtemnikov** [comment]
@teoljungberg It's ok, no problem :)

Sure, I'll add tests soon.

**Comment by teoljungberg** [comment]
Closed in favor of #50
